### PR TITLE
Deprecate --server argument to `pub publish` and `pub uploader`.

### PR DESCRIPTION
**Pull request by sigurdm**
This argument does not really work optimally.

This also fixes the error message given when
PUB_HOSTED_URL has a bad scheme.

Fixes: https://github.com/dart-lang/pub/issues/2695

### CODE REVIEW

File: lib/src/source/hosted.dart [hunk 1]
```diff
@@ -62,10 +62,10 @@ class HostedSource extends Source {
     var url = io.Platform.environment['PUB_HOSTED_URL'];
     if (url == null) return null;
     var uri = Uri.parse(url);
-    if (uri.scheme?.isEmpty ?? true) {
+    if (uri.scheme != 'http' && uri.scheme != 'https') {
```
Comment (kevmoo) [inline lib/src/source/hosted.dart:65]: `uri.hasScheme` is slightly more...uh...right

Handles weird casing, etc.

File: lib/src/command/uploader.dart [hunk 2]
```diff
@@ -28,8 +29,10 @@ class UploaderCommand extends PubCommand {
 
   UploaderCommand() {
     argParser.addOption('server',
-        defaultsTo: cache.sources.hosted.defaultUrl,
-        help: 'The package server on which the package is hosted.');
+        defaultsTo: Platform.environment['PUB_HOSTED_URL'] ??
+            'https://pub.dartlang.org',
+        help: 'The package server on which the package is hosted.\n'
+            'DEPRECATED: use `publish_to` in your pubspec.yaml or set \$PUB_HOSTED_URL.');
```
Comment (kevmoo) [inline lib/src/command/uploader.dart:35]: Be a bit more clear here `...set the $PUB_HOSTED_URL environment variable`.

File: lib/src/source/hosted.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/source/hosted.dart:65]: Hmm - I think hasScheme only tells us if we have a scheme or not.
We want to ensure that the scheme is one of 'http' and 'https'.

But the casing part is a good point. Fixed!

File: lib/src/command/uploader.dart (hunk 2 above)
Comment (sigurdm) [reply lib/src/command/uploader.dart]: Better, thanks!

File: lib/src/command/uploader.dart [hunk 3]
```diff
@@ -28,8 +29,11 @@ class UploaderCommand extends PubCommand {
 
   UploaderCommand() {
     argParser.addOption('server',
-        defaultsTo: cache.sources.hosted.defaultUrl,
-        help: 'The package server on which the package is hosted.');
+        defaultsTo: Platform.environment['PUB_HOSTED_URL'] ??
+            'https://pub.dartlang.org',
+        help: 'The package server on which the package is hosted.\n'
+            'DEPRECATED: use `publish_to` in your pubspec.yaml or set the\n'
+            '\$PUB_HOSTED_URL environment variable.');
```
Comment (jonasfj) [inline lib/src/command/uploader.dart:36]: No need to write that it's deprecated... just make it flag hidden. And then print as warning if `wasParsed('server') && isTerminal`...

File: lib/src/command/lish.dart [hunk 4]
```diff
@@ -61,7 +61,9 @@ class LishCommand extends PubCommand {
         negatable: false,
         help: 'Publish without confirmation if there are no errors.');
     argParser.addOption('server',
-        help: 'The package server to which to upload this package.');
+        help: 'The package server to which to upload this package.\n'
+            'DEPRECATED: use `publish_to` in your pubspec.yaml or set the \n'
+            '\$PUB_HOSTED_URL environment variable.');
```
Comment (jonasfj) [inline lib/src/command/lish.dart:66]: Similar here, can't we just hide the flag?

File: lib/src/command/uploader.dart (hunk 3 above)
Comment (sigurdm) [reply lib/src/command/uploader.dart]: Done

File: lib/src/command/lish.dart (hunk 4 above)
Comment (sigurdm) [reply lib/src/command/lish.dart]: Done

**Review by jonasfj — APPROVED** [review]

File: lib/src/source/hosted.dart [hunk 5]
```diff
@@ -62,10 +62,10 @@ class HostedSource extends Source {
     var url = io.Platform.environment['PUB_HOSTED_URL'];
     if (url == null) return null;
     var uri = Uri.parse(url);
-    if (uri.scheme?.isEmpty ?? true) {
+    if (!['http', 'https'].contains(uri.scheme.toLowerCase())) {
```
Comment (jonasfj) [inline lib/src/source/hosted.dart:65]: Why not `uri.hasScheme('http') || uri.hasScheme('https')`

**Review by jonasfj — APPROVED** [review]

File: lib/src/source/hosted.dart (hunk 5 above)
Comment (jonasfj) [reply lib/src/source/hosted.dart:65]: I mean: `uri.isScheme('http') || uri.isScheme('https')`

Comment (jonasfj) [reply lib/src/source/hosted.dart:65]: https://api.dart.dev/stable/2.10.3/dart-core/Uri/isScheme.html

Comment (sigurdm) [reply lib/src/source/hosted.dart:65]: Ah - that works!

Comment (sigurdm) [reply lib/src/source/hosted.dart]: Done

File: lib/src/source/hosted.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/source/hosted.dart]: Ah you mean `uri.isScheme` didn't know about that!

Fixed!

**Review by jonasfj — APPROVED** [review]

**Pull request merged**